File: AlwaysAnimate/src/main/java/com/aliucord/plugins/AlwaysAnimate.java
package com.aliucord.plugins;

import android.content.Context;
import android.widget.ImageView;

import androidx.annotation.NonNull;

import com.aliucord.entities.Plugin;
import com.aliucord.patcher.PinePatchFn;
import com.aliucord.patcher.PinePrePatchFn;
import com.aliucord.utils.ReflectUtils;
import com.discord.models.guild.Guild;
import com.discord.models.user.CoreUser;
import com.discord.utilities.icon.IconUtils;
import com.discord.widgets.chat.list.adapter.WidgetChatListAdapterItemMessage;
import com.discord.widgets.chat.list.entries.ChatListEntry;
import com.discord.widgets.chat.list.entries.MessageEntry;
import com.discord.widgets.guilds.list.GuildListViewHolder;

public class AlwaysAnimate extends Plugin {
    public AlwaysAnimate() {
        settingsTab = new SettingsTab(PluginSettings.class, SettingsTab.Type.BOTTOM_SHEET).withArgs(settings);
    }

    @NonNull
    @Override
    public Manifest getManifest() {
        Manifest manifest = new Manifest();
        manifest.authors = new Manifest.Author[]{ new Manifest.Author("Möth", 289556910426816513L) };
        manifest.description = "Allows making server icons and member avatars always animate.";
        manifest.version = "1.0.1";
        manifest.updateUrl = "https://raw.githubusercontent.com/litleck/aliucord-plugins/builds/updater.json";
        return manifest;
    }

    @Override
    public void start(Context context) throws NoSuchMethodException {
        if (settings.getBool("guildIcon", true)) {
            patcher.patch(GuildListViewHolder.GuildViewHolder.class.getDeclaredMethod("configureGuildIconImage", Guild.class, boolean.class), new PinePrePatchFn(callFrame -> callFrame.args[1] = true));
        }

        if (settings.getBool("authorAvatar", true)) {
            final IconUtils iconUtils = IconUtils.INSTANCE;
            patcher.patch(WidgetChatListAdapterItemMessage.class.getDeclaredMethod("onConfigure", int.class, ChatListEntry.class), new PinePatchFn(callFrame -> {
                MessageEntry messageEntry = (MessageEntry) callFrame.args[1];
                CoreUser coreUser = new CoreUser(messageEntry.getMessage().getAuthor());
                if (!iconUtils.isImageHashAnimated(coreUser.getAvatar())) return;

                try {
                    ImageView imageView = (ImageView) ReflectUtils.getField(callFrame.thisObject, "itemAvatar");
                    if (imageView != null) IconUtils.setIcon(imageView, coreUser.getAvatar());
                    // String avatar = iconUtils.getForGuildMember(messageEntry.getAuthor(), 1, true);
                } catch (NoSuchFieldException | IllegalAccessException e) {
                    e.printStackTrace();
                }
            }));
        }
    }

    @Override
    public void stop(Context context) { patcher.unpatchAll(); }
}

File: AlwaysAnimate/src/main/java/com/aliucord/plugins/AlwaysAnimate.kt
package com.aliucord.plugins

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.OvalShape
import android.os.PowerManager
import android.widget.ImageView
import com.aliucord.annotations.AliucordPlugin
import com.aliucord.entities.Plugin
import com.aliucord.patcher.PinePatchFn
import com.aliucord.patcher.PinePrePatchFn
import com.aliucord.plugins.alwaysanimate.PluginSettings
import com.aliucord.utils.DimenUtils
import com.discord.utilities.icon.IconUtils
import com.discord.utilities.images.MGImages
import com.facebook.drawee.view.SimpleDraweeView

@AliucordPlugin
class AlwaysAnimate : Plugin() {
    init {
        settingsTab = SettingsTab(PluginSettings::class.java, SettingsTab.Type.BOTTOM_SHEET).withArgs(settings)
    }

    override fun start(context: Context) {
        val powerManager = context.getSystemService(Context.POWER_SERVICE) as PowerManager;
        if (settings.getBool("batterySaver", true) && powerManager.isPowerSaveMode) return;

        if (settings.getBool("guildIcons", true)) {
            patcher.patch(IconUtils::class.java.getDeclaredMethod("getForGuild", Long::class.javaObjectType, String::class.javaObjectType, String::class.javaObjectType, Boolean::class.java, Int::class.javaObjectType), PinePrePatchFn {
                it.args[3] = true
            })
        }

        if (settings.getBool("avatars", true)) {
            patcher.patch(IconUtils::class.java.getDeclaredMethod("getForUser", Long::class.javaObjectType, String::class.javaObjectType, Int::class.javaObjectType, Boolean::class.javaPrimitiveType, Int::class.javaObjectType), PinePrePatchFn {
                it.args[3] = true
            })

            patcher.patch(IconUtils::class.java.getDeclaredMethod("setIcon", ImageView::class.java, String::class.java, Int::class.javaPrimitiveType, Int::class.javaPrimitiveType, Boolean::class.javaPrimitiveType, Function1::class.java, MGImages.ChangeDetector::class.java), PinePatchFn {
                val simpleDraweeView = it.args[0] as SimpleDraweeView
                simpleDraweeView.apply {
                    clipToOutline = true
                    if (settings.getBool("roundedAvatars", true)) {
                        background =
                            ShapeDrawable(OvalShape()).apply { paint.color = Color.TRANSPARENT }
                    } else {
                        background =
                            GradientDrawable().apply { shape = GradientDrawable.RECTANGLE }.apply {
                                cornerRadius = DimenUtils.dpToPx(3).toFloat()
                                setColor(Color.TRANSPARENT)
                            }
                    }
                }
            })
        }
    }

    override fun stop(context: Context) = patcher.unpatchAll()
}